### ISSUE TITLE
Provide a way to safe decode collection items

### ISSUE THREAD

**Issue by JanC** [kean/CreateAPI#18]
hey,

some other open api generator tools provide a way to generate code that safely decodes elements of an array and ignores the ones that don't. This can be used in scenarios where new types are likely to be introduced in the API and yet we don't want older api clients not supporting the new types to break.

For instance, [Swaggen](https://github.com/yonaskolb/SwagGen) decodes all array using the [`decodeArray`](https://github.com/yonaskolb/SwagGen/blob/master/Templates/Swift/Sources/Coding.swift#L125-L152) which ignore elements that fail to decode.

This works however having such a global setting which simply ignores elements of all arrays is a big assumption that should not be part of the code generation library. There are cases where we want some types to be decoded safely and some case where we want them to be more strict.

I was wondering if and how CreateAPI could provide a similar functionality while remaining flexible enough.

I think It should be up to the user of CreateAPI to 
- Decide what are the collections that can have elements that fail to decode.
- What to do if an element fails to decode.

For this purpose, in our **non generated** code, we use a custom type to wrap the entities that are likely to fail. For instance, the following `Throwable` simply logs the decoding error

```swift
public struct Throwable<T: Decodable>: Decodable {
    let result: Result<T, ThrowableDecodableError>

    public init(from decoder: Decoder) throws {
        do {
            result = .success(try T(from: decoder))
        } catch let error {
            logger.info("Failed to decode: \(error)")
            result = .failure(ThrowableDecodableError())
        }
    }

    public func get() -> T? {
        return try? result.get()
    }
}
```

We then declare the collections properties like this

```swift
public struct Store: Decodable {
    public let pets: [Throwable<Pet>]
}
```

Introducing the above `Throwable` in CreateAPI would be again too much assumption and that's not what I'm proposing here. It's just for illustration.

However, I was thinking if it would make sense to create a setting such as `arrayTypeWrappers` 

```yaml
  arrayTypeWrappers: [
    "MyEntity": "MySettingType"
  ]
```
which would be a mapping of types that CreateAPI would use to simply output `[MySettingType<MyEntity>]` instead of 
 `[MyEntity]`.

So back to my `Throwable` example, a setting 
```yaml
entities:
  arrayTypeWrappers: [
    "Pet": "Throwable"
  ]
```

would generate 

```swift

public struct Store: Decodable {
    public let pets: [Throwable<Pet>]
}

public struct World: Decodable {
    public let stores: [Store] // nothing new here
}
```

Of course, the user of CreateAPI would have to provide the `Throwable` implementation for the code to compile

Any thoughts?

**Comment by kean** [comment]
Hey, @JanC. Yeah, I have this feature on a list (which I'll put on a public board in the future). 

I think ideally "safe" decoding should be built-in. The user should have an option to enable it (disabled by default) and provide an `include` and `exclude` list to customize which properties to decode safely.

```swift
public struct Store: Decodable {
    public let pets: [Pet]

    public init(from decoder: Decoder) {
         // decode pets safely (if enabled for this property)
         // can use `Throwable` type internally
    }
}
```

I'm not sure if you need this level of generalization – wrapping responses in particular types – but I think it would be nice to also provide an option to either "hide" safe decoding, or use `Throwable` in public API. If the user decides to use `Throwable`, add it to the generated code – at least that's how I approached other similar features. I don't mind also adding support for custom wrappers if you already have a `Throwable` type of its analog.

**Comment by JanC** [comment]
yes, a built-in and easy to use option for safe decoding would be nice.

I think however it should be somehow possible to expose the failed decoded entities to allow non silent failures. So as you said probably by exposing a `Throwable` in  the public generated API
This `Throwable` should keep  and expose the actual decoding error so that it can be processed further the api user.

As an example, what we currently do in our non generated `Throwable` is that we group the decoding failures and batch log them to a remote server.

**Comment by kean** [comment]
I'm for providing all of these options so that the users could decide what works best for them 👍

**Comment by JanC** [comment]
One thing came to my mind when thinking about the `arrayTypeWrappers`. As the type supplied by the api user will likely be part of another module, I think the `GenerateOptions.Entities` options should have the same `import` option as the `GenerateOptions.Paths`.

**Comment by liamnichols** [comment]
How about leveraging property wrappers for this? That would give users the ability to introduce a wide range of customisations that go beyond just arrays too

For example:

```yml
entities:
  propertyWrappers:
    Array: LossyArray
  imports:
  - MyUtilsLibWithLossyArrayPropertyWrapper
```

This would cause entities with the above type mapping to append property wrappers like the following:

```swift
// Generated by Create API
// https://github.com/CreateAPI/CreateAPI
//
// swiftlint:disable all

import Foundation
import MyUtilsLibWithLossyArrayPropertyWrapper

/// A pet title
///
/// A pet description
public struct Pet: Codable {
    public var id: Int
    /// Example: "Buddy"
    public var name: String
    public var tag: String?

    public init(id: Int, name: String, tag: String? = nil) {
        self.id = id
        self.name = name
        self.tag = tag
    }
}

public struct Store: Codable {
    @LossyArray
    public var pets: [Pet]

    public init(pets: [Pet]) {
        self.pets = pets
    }
}
```

In the above example, the property wrapper is something like this:

```swift
private struct AnyDecodableValue: Decodable {}

@propertyWrapper struct LossyArray<Element> {
    var wrappedValue: Array<Element>
}

extension LossyArray: Decodable where Element: Decodable {
    init(from decoder: Decoder) throws {
        var container = try decoder.unkeyedContainer()

        var items: [Element] = []
        while !container.isAtEnd {
            do {
                let item = try container.decode(Element.self)
                items.append(item)
            } catch {
                try container.decode(AnyDecodableValue.self) // Must decode successfully in order to move on
            }
        }

        wrappedValue = items
    }
}

extension LossyArray: Encodable where Element: Encodable {
    func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        try container.encode(wrappedValue)
    }
}
```

I haven't thought much about other ways that property wrappers could be used, but depending on that, there might be better ways to define the configuration to give more customisation.

**Comment by JanC** [comment]
Hey @liamnichols 
I think that would work as well and it's an alternative way of doing the same thing. So I believe both options would be useful for the user. Some might prefer to wrap types in generics, others in the property wrappers. Regardless of what the actual property wrapper or generic class actually does.

For your proposed suggested yaml, I think it should however be done on a specific type and not for all Arrays of any type

So for example:
```yamls
entities:
  propertyWrappers:
    Pet: LossyArray
  imports:
  - MyUtilsLibWithLossyArrayPropertyWrapper
  - ```

that way, only properties of type `[Pet]` would use the `@LossyArray`
